### PR TITLE
Add support for the `windowingBehavior` setting

### LINKED ISSUE #2227: Feature Request: Option to only allow one instance
# Description of the new feature/enhancement

When I work in the terminal (currently mainly CMDer/ConEmu) I like to keep one terminal going with multiple tabs.  
When opening a new terminal window, my terminal does not create a new instance of the application but rather opens a new tab in the currently running instance.

Best approach would probably be to make it an option, I.E., not forcing the functionality on people, but rather have it toggle-able for people like me!

# Proposed technical implementation details (optional)

Open a terminal while there is a terminal already running opens the running terminal application and creates a new terminal tab, alt. opens the terminal and allows the user to open a new tab manually.

_(Love the terminal, it's really neat!)_

### PR DESCRIPTION
**Pull request by zadjii-msft**
Adds support for the `windowingBehavior` global setting. This setting
controls how mutiple instances of `wt` behave in the absence of the `-w`
parameter. This setting has three values:
* `"useNew"`: (default) Multiple `wt` invocations (without the `-w`
  param) always create new windows. 
* `"useAnyExisting"`: When starting a new `wt`, we'll instead default to
  any existing windows. `wt -w -1` will still create new windows. 
* `"useExisting"`: Similar to `useAnyExisting`, but limits to
  windows on the current desktop. 

The IVirtualDesktopManager interface is _very_ limited. Hence why we
have to track the HWNDs manually, and ask if they're on the current
desktop. 

## Validation Steps Performed
I've been playing with it for a week now. 

References #5000
References projects/5
References #8898
Spec'd in #8135
Closes #2227
Closes https://github.com/microsoft/terminal/projects/5#card-51431448
Closes https://github.com/microsoft/terminal/projects/5#card-51431433

### CODE REVIEW

File: src/cascadia/TerminalApp/AppLogic.cpp [hunk 1]
```diff
@@ -1194,27 +1195,44 @@ namespace winrt::TerminalApp::implementation
...
+                {
+                case WindowingMode::UseNew:
+                    return WindowingBehaviorUseNew;
+                case WindowingMode::UseExistingSameDesktop:
+                    return WindowingBehaviorUseExistingSameDesktop;
+                case WindowingMode::UseExisting:
+                    return WindowingBehaviorUseExisting;
+                default:
+                    return WindowingBehaviorUseNew;
+                }
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/AppLogic.cpp:1234]: Random idea (this may be awful). Since enums are just numbers, could we have `static_cast<int>(WindowingMode::UseNew)` == `WindowingBehaviorUseNew` (and the same for the rest)?

File: src/cascadia/Remoting/WindowManager.h [hunk 2]
```diff
@@ -59,6 +59,7 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
         bool _performElection();
         void _createPeasantThread();
         void _waitOnMonarchThread();
+
```
Comment (carlos-zamora) [inline src/cascadia/Remoting/WindowManager.h:62]: nit: could undo this change

File: src/cascadia/Remoting/Peasant.idl [hunk 3]
```diff
@@ -11,12 +11,15 @@ namespace Microsoft.Terminal.Remoting
 
         String[] Commandline { get; set; };
         String CurrentDirectory();
+
```
Comment (carlos-zamora) [inline src/cascadia/Remoting/Peasant.idl:14]: nit: you can undo this change

File: src/cascadia/Remoting/Monarch.cpp [hunk 4]
```diff
@@ -20,6 +21,11 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
     Monarch::Monarch() :
         _ourPID{ GetCurrentProcessId() }
     {
+        try
+        {
+            _desktopManager = winrt::create_instance<IVirtualDesktopManager>(__uuidof(VirtualDesktopManager));
```
Comment (carlos-zamora) [inline src/cascadia/Remoting/Monarch.cpp:26]: Do we need a new one each time a monarch is created? Or should we make `_desktopManager` static, and only create it if we don't have one?

File: src/cascadia/Remoting/Monarch.cpp [hunk 5]
```diff
@@ -130,63 +139,239 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
...
 
-        // We haven't yet been told the MRU peasant. Just use the first one.
-        // This is just gonna be a random one, but really shouldn't happen
-        // in practice. The WindowManager should set the MRU peasant
-        // immediately as soon as it creates the monarch/peasant for the
-        // first window.
-        if (_peasants.size() > 0)
+        // Here, there's at least one MRU peasant.
+        //
+        // We're going to iterate over these peasants toll we find one that both:
```
Comment (carlos-zamora) [inline src/cascadia/Remoting/Monarch.cpp:291]:
```suggestion
        // We're going to iterate over these peasants until we find one that both:
```

**Review by carlos-zamora — CHANGES_REQUESTED** [review]
Make sure you add this to the docs, schema, and settings UI.

Settings UI should be super simple now that SettingContainer is in. Just a matter of...
- setting up the EnumMapping
- initialize the EnumEntry in the SUI page you want
- set up a SettingContainer with it

That said, I understand if this is a separate PR.

Really just blocking over the few settings stuff. Implementation looks solid.

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 6]
```diff
@@ -556,12 +556,30 @@ winrt::fire_and_forget AppHost::_WindowActivated()
...
         peasant.ActivateWindow(args);
     }
 }
+
+GUID AppHost::_CurrentDesktopGuid()
+{
+    GUID currentDesktopGuid{ 0 };
+    try
+    {
+        auto manager = winrt::create_instance<IVirtualDesktopManager>(__uuidof(VirtualDesktopManager));
```
[the remark this answers is not in the document]
Comment (DHowett) [reply src/cascadia/WindowsTerminal/AppHost.cpp:577]: Prefer not using const auto& for references even though it does work the way you expect. It reads as incorrect, so we should optimize for readability.  const auto without the reference should suffice.

File: src/inc/WindowingBehavior.h [hunk 7]
```diff
@@ -0,0 +1,10 @@
+/*++
+Copyright (c) Microsoft Corporation
+Licensed under the MIT license.
+--*/
+
+#pragma once
+constexpr int64_t WindowingBehaviorUseCurrent{ 0 };
+constexpr int64_t WindowingBehaviorUseNew{ -1 };
+constexpr int64_t WindowingBehaviorUseExistingSameDesktop{ -2 };
+constexpr int64_t WindowingBehaviorUseExisting{ -3 };
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/inc/WindowingBehavior.h:10]: What would static add to these? Wouldn't that actually _hide_ these definitions from the things consuming them (since they're in other files)?

File: src/cascadia/Remoting/Monarch.cpp (hunk 4 above)
Comment (zadjii-msft) [reply src/cascadia/Remoting/Monarch.cpp:26]: Yea, we do, because there's only ever one Monarch pre-process. So the next process can't really re-use the one from the previous process 😉

File: src/cascadia/TerminalApp/AppLogic.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/AppLogic.cpp:1234]: Meh, theoretically yes, but I didn't want to rely on that behavior necessarily. I'm already going to have to have the constants somewhere (since I don't want to have to pull in all of TSM to Remoting just for some constants) so I'm gonna have to do this either here or in Remoting

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw [hunk 8]
```diff
@@ -232,6 +232,21 @@
...
     <value>Maximized</value>
   </data>
+  <data name="Globals_WindowingBehavior.Header" xml:space="preserve">
+    <value>Windowing behavior</value>
+  </data>
+  <data name="Globals_WindowingBehavior.HelpText" xml:space="preserve">
+    <value>Controls how new terminal instances attach to existing windows</value>
+  </data>
+  <data name="Globals_WindowingBehaviorUseNew.Content" xml:space="preserve">
+    <value>Create a new window</value>
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:242]: I'd suggest throwing in an "Always" here.

File: doc/cascadia/profiles.schema.json [hunk 9]
```diff
@@ -812,6 +812,16 @@
...
             }
           ]
+        },
+        "windowingBehavior": {
+          "default": "useNew",
+          "description": "Controls how new terminal instances attach to existing windows.",
+          "enum": [
+            "useNew",
+            "useAny",
+            "useAnyExisting"
```
Comment (DHowett) [inline doc/cascadia/profiles.schema.json:822]: these two string values aren't as clear as their descriptions; should we make them a little more self-evident?

File: src/cascadia/Remoting/Monarch.cpp [hunk 10]
```diff
@@ -130,63 +139,239 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
...
-        // when the user executes a `wt -w 0`, we won't be able to find that
-        // peasant, and it'll open in a new window instead of the current one.
-        if (args.ActivatedTime() > _lastActivatedTime)
+        // Start by making a local copy of these args. It's easier for us if our
+        // tracking of these args is all in-proc. That way, the only thing that
+        // could fail due to the peasant dying is _this first copy_.
+        winrt::com_ptr<implementation::WindowActivatedArgs> localArgs{ nullptr };
+        try
+        {
+            localArgs = winrt::make_self<implementation::WindowActivatedArgs>(args.PeasantID(),
```
Comment (DHowett) [inline src/cascadia/Remoting/Monarch.cpp:174]: nit: I'd expect this to act more like a copy constructor

Comment (DHowett) [reply src/cascadia/Remoting/Monarch.cpp:174]: (take in a projected type, unbox it internally. that way, it's a one-stop shop for "where do I add new fields to copy?")

File: src/cascadia/Remoting/Monarch.cpp [hunk 11]
```diff
@@ -130,63 +139,239 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
...
+    //   activated (so that we don't leave an old entry for it in the list).
+    // Arguments:
+    // - peasantID: The ID of the peasant to remove from the MRU list
+    // Return Value:
+    // - <none>
+    void Monarch::_clearOldMruEntries(const uint64_t peasantID)
+    {
+        auto result = std::find_if(_mruPeasants.begin(),
+                                   _mruPeasants.end(),
+                                   [peasantID](auto other) {
```
Comment (DHowett) [inline src/cascadia/Remoting/Monarch.cpp:201]: nit: `auto&&` will let the compiler decide what kind of parameter type is best. `auto` requires a copy (and the addref/release that you'd get by doing that)

File: src/cascadia/Remoting/Monarch.cpp [hunk 12]
```diff
@@ -130,63 +139,239 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
...
+        // If the peasant is dead, then we'll remove it, and try the next one.
+        // Once we find one that's alive, we'll either:
+        // * check if we only want a peasant on the current desktop, and if so,
+        //   check if this peasant is on the current desktop.
+        //   - If it isn't on the current desktop, we'll loop again, on the
+        //     following peasant.
+        // * If we don't care, then we'll just return that one.
+
+        int positionInList = 0;
+        while (_mruPeasants.begin() + positionInList < _mruPeasants.end())
```
Comment (DHowett) [inline src/cascadia/Remoting/Monarch.cpp:304]: nit: a little note in the comment as to why an iterator is not suitable would go a long way :smile:

Comment (DHowett) [reply src/cascadia/Remoting/Monarch.cpp:304]: nit: can these be `cbegin` and `cend`? we're not mutating through the iterators

Comment (DHowett) [reply src/cascadia/Remoting/Monarch.cpp:304]: (and below)

File: src/cascadia/Remoting/Monarch.cpp [hunk 13]
```diff
@@ -130,63 +139,239 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
...
+        //
+        // If the peasant is dead, then we'll remove it, and try the next one.
+        // Once we find one that's alive, we'll either:
+        // * check if we only want a peasant on the current desktop, and if so,
+        //   check if this peasant is on the current desktop.
+        //   - If it isn't on the current desktop, we'll loop again, on the
+        //     following peasant.
+        // * If we don't care, then we'll just return that one.
+
+        int positionInList = 0;
```
Comment (DHowett) [inline src/cascadia/Remoting/Monarch.cpp:303]: nit: `i` is probably _fine_ :P

File: src/cascadia/Remoting/Monarch.cpp [hunk 14]
```diff
@@ -130,63 +139,239 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
...
-                // first peasant should also _be us_. So we should be able to
-                // get our own ID.
-                return 0;
+                // Check if this peasant is actually on this desktop. We can't
+                // simply get the GUID of the current desktop. We have to ask if
+                // the HWND is on the current desktop.
+                BOOL onCurrentDesktop{ false };
+
+                // SUCCEEDED_LOG will log if it failed, and return true if it SUCCEEDED
+                if (SUCCEEDED_LOG(_desktopManager->IsWindowOnCurrentVirtualDesktop((HWND)mruWindowArgs.Hwnd(),
```
Comment (DHowett) [inline src/cascadia/Remoting/Monarch.cpp:331]: prefer a more expressive cast here, like `static_cast<HWND>`

```suggestion
                if (SUCCEEDED_LOG(_desktopManager->IsWindowOnCurrentVirtualDesktop(static_cast<HWND>(mruWindowArgs.Hwnd()),
```

File: src/cascadia/Remoting/Peasant.idl [hunk 15]
```diff
@@ -11,12 +11,15 @@ namespace Microsoft.Terminal.Remoting
 
         String[] Commandline { get; set; };
         String CurrentDirectory();
+
     };
 
     runtimeclass WindowActivatedArgs
     {
         WindowActivatedArgs(UInt64 peasantID, Guid desktopID, Windows.Foundation.DateTime activatedTime);
+        WindowActivatedArgs(UInt64 peasantID, UInt64 hwnd, Guid desktopID, Windows.Foundation.DateTime activatedTime);
```
Comment (DHowett) [inline src/cascadia/Remoting/Peasant.idl:20]: Interesting -- is this needed on the public surface? I only saw it used via `make_self`.

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw [hunk 16]
```diff
@@ -232,6 +232,21 @@
   <data name="Globals_LaunchModeMaximized.Content" xml:space="preserve">
     <value>Maximized</value>
   </data>
+  <data name="Globals_WindowingBehavior.Header" xml:space="preserve">
+    <value>Windowing behavior</value>
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:236]: Hmm. Should we workshop the phrasing?

```
New instances of Terminal open in...
... a new window
... the most recently-used window on the same desktop
... the most recently-used window on any desktop
```

How do other applications with automatic glomming phrase this? This seems a little too "technical" for the SUI

File: src/inc/WindowingBehavior.h [hunk 17]
```diff
@@ -0,0 +1,10 @@
+/*++
+Copyright (c) Microsoft Corporation
+Licensed under the MIT license.
+--*/
+
+#pragma once
+constexpr int64_t WindowingBehaviorUseCurrent{ 0 };
```
Comment (DHowett) [inline src/inc/WindowingBehavior.h:7]: do we have a `cascadia/inc`?

Comment (DHowett) [inline src/inc/WindowingBehavior.h:7]: I don't honestly know how these symbols work when included in multiple translation units in the same library. Carlos' question about static does stick in my mind.

Comment (DHowett) [reply src/inc/WindowingBehavior.h:7]: this seems like a contract between remoting and apphost honestly

File: src/cascadia/Remoting/Peasant.idl (hunk 15 above)
Comment (DHowett) [reply src/cascadia/Remoting/Peasant.idl:20]: ONCE AGAIN ignore me haha

**Review by DHowett — CHANGES_REQUESTED** [review]
26/28

File: src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp [hunk 18]
```diff
@@ -308,6 +309,8 @@ void GlobalAppSettings::LayerJson(const Json::Value& json)
 
     JsonUtils::GetValueForKey(json, FocusFollowMouseKey, _FocusFollowMouse);
 
+    JsonUtils::GetValueForKey(json, WindowingBehaviorKey, _WindowingBehavior);
```
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:312]: > @carlos-zamora why the heck doesn't this work? It works to change the value of the setting, but it does _not_ work on the initial load of the Launch page. It always selects UseNew. But if I change the value, DONT SAVE, navigate away from the page, then back to the page, then it loads correctly. What gives?

@zadjii-msft you need to update "GlobalAppSettings::Copy()". When SUI is created, we clone the settings model using `Copy`. So your value isn't getting passed over, resulting in the default value being used. After that, SUI should work just fine.

Surprised to hear about the "navigate away" thing. I couldn't get a repro on that, this should fix your issue entirely.

File: doc/cascadia/profiles.schema.json (hunk 9 above)
Comment (zadjii-msft) [reply doc/cascadia/profiles.schema.json:822]: Whoops, yea I changed those before I opened Teams to  to check what we had settled on 😋

File: src/inc/WindowingBehavior.h (hunk 17 above)
Comment (zadjii-msft) [reply src/inc/WindowingBehavior.h:7]: okay sure;

Yea this one's awkward because TerminalApp definitely needs to be aware of these values for `FindTargetWindow` to work right, but App doesn't otherwise depend on remoting at all. Plus, we can't just have `FindTargetWindow` return a `WindowingMode`, because it technically needs to return a `union{ WindowingMode, uint }`, so ¯\\\_(ツ)_/¯

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 16 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:236]: I couldn't find anyone else with a setting for this. That being said, after the Tabbed Sets fiasco, I'm sure that it's a needed setting, so let's go with the proposal here. I like it.

File: src/cascadia/Remoting/Monarch.cpp (hunk 14 above)
Comment (zadjii-msft) [reply src/cascadia/Remoting/Monarch.cpp:331]: Oh, it does _not_ like that, I think because an HWND is technically a `void*`, so this would need to be a `reinterpret_cast`

File: src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:312]: Oh yep. derp.

File: src/inc/WindowingBehavior.h (hunk 17 above)
Comment (zadjii-msft) [reply src/inc/WindowingBehavior.h:7]: https://stackoverflow.com/a/50489130

I don't know if that answers the question but it sounds like I'm doing it right?

File: src/inc/WindowingBehavior.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/inc/WindowingBehavior.h]: So it turns out, the `static` is implied by `constexpr` in a header: https://stackoverflow.com/a/50489130

**Comment by zadjii-msft** [comment]
Now the setting works right in the UI as well:
image

File: doc/cascadia/profiles.schema.json [hunk 19]
```diff
@@ -812,6 +812,16 @@
               "type": "string"
             }
           ]
+        },
+        "windowingBehavior": {
+          "default": "useNew",
+          "description": "Controls how new terminal instances attach to existing windows. \"useNew\" will create a new window always. \"useExisting\" will create new tabs in the most recently used window on this desktop, and \"useAnyExisting\" will create tabs in the most recent window on any desktop.",
```
Comment (carlos-zamora) [inline doc/cascadia/profiles.schema.json:818]:
```suggestion
          "description": "Controls how new terminal instances attach to existing windows. \"useNew\" will always create a new window. \"useExisting\" will create new tabs in the most recently used window on this desktop, and \"useAnyExisting\" will create tabs in the most recent window on any desktop.",
```

Comment (carlos-zamora) [reply doc/cascadia/profiles.schema.json:818]: Should we also replace "desktop" with "virtual desktop"?

File: src/cascadia/TerminalApp/AppLogic.cpp [hunk 20]
```diff
@@ -1194,27 +1195,44 @@ namespace winrt::TerminalApp::implementation
...
+                switch (windowingBehavior)
+                {
+                case WindowingMode::UseNew:
+                    return WindowingBehaviorUseNew;
+                case WindowingMode::UseExisting:
+                    return WindowingBehaviorUseExisting;
+                case WindowingMode::UseAnyExisting:
+                    return WindowingBehaviorUseAnyExisting;
+                default:
+                    return WindowingBehaviorUseNew;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/AppLogic.cpp:1233]:
```suggestion
                case WindowingMode::UseExisting:
                    return WindowingBehaviorUseExisting;
                case WindowingMode::UseAnyExisting:
                    return WindowingBehaviorUseAnyExisting;
                case WindowingMode::UseNew:
                default:
                    return WindowingBehaviorUseNew;
```
nit

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw [hunk 21]
```diff
@@ -232,6 +232,21 @@
...
+  </data>
+  <data name="Globals_WindowingBehaviorUseNew.Content" xml:space="preserve">
+    <value>a new window</value>
+  </data>
+  <data name="Globals_WindowingBehaviorUseAnyExisting.Content" xml:space="preserve">
+    <value>the most recently used window</value>
+  </data>
+  <data name="Globals_WindowingBehaviorUseExisting.Content" xml:space="preserve">
+    <value>the most recently used window on this desktop</value>
+  </data>
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:249]: @cinnamon-msft please take a look here. Easier to get the wording correct now than have to get it localized and change it later.

**Review by carlos-zamora — APPROVED** [review]
This is great! Just tagged Kayla on one thing because I don't want us to have to revisit the wordsmithing conversation later.

**Comment by DHowett** [comment]
minor minor nit: PR body needs updating to reflect current enum names for things

**Comment by DHowett** [comment]
**DO NOT AUTOMERGE**

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw [hunk 22]
```diff
@@ -232,6 +232,22 @@
   <data name="Globals_LaunchModeMaximized.Content" xml:space="preserve">
     <value>Maximized</value>
   </data>
+  <data name="Globals_WindowingBehavior.Header" xml:space="preserve">
+    <value>New windows open in...</value>
```
Comment (cinnamon-msft) [inline src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:236]:
```suggestion
    <value>wt.exe window behavior</value>
```

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw [hunk 23]
```diff
@@ -232,6 +232,22 @@
...
   </data>
+  <data name="Globals_WindowingBehavior.Header" xml:space="preserve">
+    <value>New windows open in...</value>
+    <comment>This text is a fragment of a sentence that indicates the options below it, in a user interface. It is a header, but it should be read like a sentence that ends with one of {Globals_WindowingBehaviorUseNew.Content}, {Globals_WindowingBehaviorUseAnyExisting.Content}, or {Globals_WindowingBehaviorUseExisting.Content}.</comment>
+  </data>
+  <data name="Globals_WindowingBehavior.HelpText" xml:space="preserve">
+    <value>Controls how new terminal instances attach to existing windows.</value>
+  </data>
+  <data name="Globals_WindowingBehaviorUseNew.Content" xml:space="preserve">
+    <value>a new window</value>
```
Comment (cinnamon-msft) [inline src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:243]:
```suggestion
    <value>Create a new window</value>
```

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw [hunk 24]
```diff
@@ -232,6 +232,22 @@
...
+  </data>
+  <data name="Globals_WindowingBehaviorUseAnyExisting.Content" xml:space="preserve">
+    <value>the most recently used window</value>
```
Comment (cinnamon-msft) [inline src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:246]:
```suggestion
    <value>Attach to the most recently used window</value>
```

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw [hunk 25]
```diff
@@ -232,6 +232,22 @@
...
+  </data>
+  <data name="Globals_WindowingBehaviorUseExisting.Content" xml:space="preserve">
+    <value>the most recently used window on this desktop</value>
```
Comment (cinnamon-msft) [inline src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:249]:
```suggestion
    <value>Attach to the most recently used window on this desktop</value>
```

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw [hunk 26]
```diff
@@ -232,6 +232,22 @@
   <data name="Globals_LaunchModeMaximized.Content" xml:space="preserve">
     <value>Maximized</value>
   </data>
+  <data name="Globals_WindowingBehavior.Header" xml:space="preserve">
+    <value>New windows open in...</value>
+    <comment>This text is a fragment of a sentence that indicates the options below it, in a user interface. It is a header, but it should be read like a sentence that ends with one of {Globals_WindowingBehaviorUseNew.Content}, {Globals_WindowingBehaviorUseAnyExisting.Content}, or {Globals_WindowingBehaviorUseExisting.Content}.</comment>
+  </data>
+  <data name="Globals_WindowingBehavior.HelpText" xml:space="preserve">
+    <value>Controls how new terminal instances attach to existing windows.</value>
```
Comment (cinnamon-msft) [inline src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:240]:
```suggestion
    <value>Controls how new terminal instances launched from wt.exe attach to existing windows.</value>
```

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 22 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:236]: Hmmmmm... this will be weird when Def Term is in

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 26 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:240]: This feels a little redundant though, right? Like, _all_ the windows are ones spawned from `wt.exe`. Do we really need to specify that?

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 22 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:236]: @cinnamon-msft Yea I concur. Plus I'm just not sure that the users will always think "ah yes, I'm running `wt.exe`". Like when running from the start menu - that's _also_ going to behave this way. 

I'm fine flipping these back since what's proposed is basically what I had originally - I just want to make sure we're _sure_, and I'm not so sure about mentioning "wt.exe"

**Comment by DHowett** [comment]
(I personally love the sentenceized version of it ("new windows open in..."))

**Comment by DHowett** [comment]
Demo gif moved from main body

windowingBehavior-useExisting-000

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 22 above)
Comment (cinnamon-msft) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:236]: Maybe `Launch window behavior`?

**Review by DHowett — APPROVED** [review]
LOVE THIS.

**Comment by DHowett** [comment]
I guess I am incredibly dull and got too excited AND AUTOMERGED THE "DO NOT AUTOMERGE" PR.
